pydoctor/epydoc2stan.py

```diff
@@ -552,8 +561,93 @@ def __init__(self, annotation: ast.expr) -> None:
         self.annotation = annotation
 
     def to_stan(self, docstring_linker: DocstringLinker) -> Tag:
-        src = astor.to_source(self.annotation).strip()
-        ret: Tag = tags.code(src)
+        tag: Tag = tags.code
+        tag(_AnnotationFormatter(docstring_linker).visit(self.annotation))
+        return tag
+
+
+class _AnnotationFormatter(ast.NodeVisitor):
+
+    def __init__(self, linker: DocstringLinker):
+        super().__init__()
+        self.linker = linker
+
+    def _handle_name(self, identifier: str) -> Tag:
+        url = self.linker.resolve_identifier(identifier)
+
+        tag: Tag
+        if url is None:
+            tag = tags.transparent(identifier)
+        else:
+            tag = tags.a(identifier, href=url, class_='code')
+        return tag
+
+    def _handle_constant(self, node: ast.expr, value: object) -> Tag:
+        if value in (False, True, None, NotImplemented):
+            # Link built-in constants to the standard library.
+            # Ellipsis is not included here, both because its code syntax is
+            # different from its constant's name and because its documentation
+            # is not relevant to annotations.
+            return self._handle_name(str(value))
+        else:
+            return self.generic_visit(node)
+
+    def _handle_sequence(self, tag: Tag, sequence: Iterable[ast.expr]) -> None:
+        first = True
+        for elem in sequence:
+            if first:
+                first = False
+            else:
+                tag(', ')
+            tag(self.visit(elem))
+
+    def visit_Name(self, node: ast.Name) -> Tag:
+        return self._handle_name(node.id)
+
+    def visit_Attribute(self, node: ast.Attribute) -> Tag:
+        parts = []
+        curr: ast.expr = node
+        while isinstance(curr, ast.Attribute):
+            parts.append(curr.attr)
+            curr = curr.value
+        if not isinstance(curr, ast.Name):
+            return self.generic_visit(node)
```

Comment:
missing coverage?

Reply:
Yes. The expected pattern for attribute lookups is that the sequence ends with a name, but the types from the AST module don't guarantee that. I'll try to find a code fragment that triggers this early return and add it to the tests.

Reply:
I can't actually find any code fragment that leads to this line being executed. While `ast.Attribute.value` is documented as having type `ast.expr`, in practice there doesn't seem to be any syntactically valid Python that produces an `ast.Attribute` for which the `value` is neither an `ast.Name` nor another `ast.Attribute`. 

However, if such an expression would exist, perhaps in a future Python version with new grammatical options, I think we should have pydoctor handle it gracefully rather than crashing on it. Note however that "handle it gracefully" currently means "let `astor` deal with it", but `astor` won't be able to format future grammar either without updates to its code.

I'll leave it as-is for now, but this is something we might want to revisit at some point. For example if we want to drop the dependency on `astor`: now that we do custom AST formatting, we don't actually use all that much functionality from that library anymore.